batch/batch/front_end/front_end.py

```diff
@@ -338,52 +338,67 @@ async def _get_job_log_from_record(app, batch_id, job_id, record):
 
-        async def _read_log_from_gcs(task):
-            try:
-                data = await file_store.read_log_file(
-                    batch_format_version, batch_id, job_id, record['attempt_id'], task
-                )
-            except FileNotFoundError:
-                id = (batch_id, job_id)
-                log.exception(f'missing log file for {id} and task {task}')
-                data = 'ERROR: could not find log file'
-            return task, data
+    if state == 'Cancelled' and record['last_cancelled_attempt_id'] is None:
+        return None
 
-        spec = json.loads(record['spec'])
-        tasks = []
+    assert state in ('Error', 'Failed', 'Success', 'Cancelled')
 
-        has_input_files = batch_format_version.get_spec_has_input_files(spec)
-        if has_input_files:
-            tasks.append('input')
+    batch_format_version = BatchFormatVersion(record['format_version'])
+    attempt_id = record['attempt_id'] or record['last_cancelled_attempt_id']
+    assert attempt_id
 
-        tasks.append('main')
+    file_store: FileStore = app['file_store']
+    batch_format_version = BatchFormatVersion(record['format_version'])
 
-        has_output_files = batch_format_version.get_spec_has_output_files(spec)
-        if has_output_files:
-            tasks.append('output')
+    async def _read_log_from_gcs(task):
+        try:
+            data = await file_store.read_log_file(batch_format_version, batch_id, job_id, attempt_id, task)
+        except FileNotFoundError:
+            id = (batch_id, job_id)
+            log.exception(f'missing log file for {id} and task {task}')
+            data = 'ERROR: could not find log file'
+        return task, data
+
+    spec = json.loads(record['spec'])
+    tasks = []
+
+    has_input_files = batch_format_version.get_spec_has_input_files(spec)
+    if has_input_files:
+        tasks.append('input')
 
-        return dict(await asyncio.gather(*[_read_log_from_gcs(task) for task in tasks]))
+    tasks.append('main')
 
-    return None
+    has_output_files = batch_format_version.get_spec_has_output_files(spec)
+    if has_output_files:
+        tasks.append('output')
+
+    return dict(await asyncio.gather(*[_read_log_from_gcs(task) for task in tasks]))
 
 
 async def _get_job_log(app, batch_id, job_id):
     db: Database = app['db']
 
     record = await db.select_and_fetchone(
         '''
-SELECT jobs.state, jobs.spec, ip_address, format_version, jobs.attempt_id
+SELECT jobs.state, jobs.spec, ip_address, format_version, jobs.attempt_id, t.attempt_id AS last_cancelled_attempt_id
 FROM jobs
 INNER JOIN batches
   ON jobs.batch_id = batches.id
 LEFT JOIN attempts
   ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id AND jobs.attempt_id = attempts.attempt_id
 LEFT JOIN instances
   ON attempts.instance_name = instances.name
+LEFT JOIN (
+  SELECT batch_id, job_id, attempt_id
+  FROM attempts
+  WHERE reason = "cancelled"
+  ORDER BY end_time DESC
+  LIMIT 1
```

Comment:
This doesn't seem right to me.
```
select batch_id, job_id, attempt_id
from attempts
where reason = "cancelled"
order by end_time desc
limit 1
```

Will return a table with exactly one row, the most recently completed, cancelled attempt. There's no guarantee that that job will be for the same job nor the same batch that we need.

I think you want this query:
```
select batch_id, job_id, attempt_id
from attempts
where reason = "cancelled" and batch_id = %s, job_id = %s
order by end_time desc
limit 1
```

That should be rather efficient too.